compiler/noirc_frontend/src/monomorphization/debug_types.rs

```diff
@@ -0,0 +1,70 @@
+use crate::hir_def::types::Type;
+pub use noirc_errors::debug_info::{Types, VariableTypes, Variables};
+use noirc_printable_type::PrintableType;
+use std::collections::HashMap;
+
+/// We keep a collection of the debug variables and their types in this
+/// structure. The fe_var_id refers to the ID given when inserting the
+/// instrumentation probe. This variable does not have a type yet and hence it
+/// can be instantiated multiple types if it's in the context of a generic
+/// variable. The var_id refers to the ID of the instantiated variable which
+/// will have a valid type.
+#[derive(Debug, Clone, Default)]
+pub struct DebugTypes {
+    fe_to_vars: HashMap<u32, u32>,          // fe_var_id => var_id
+    variables: HashMap<u32, (String, u32)>, // var_id => (var_name, type_id)
+    types: HashMap<PrintableType, u32>,
+    id_to_type: HashMap<u32, PrintableType>,
+    next_type_id: u32,
+    next_var_id: u32,
+}
```

Comment:
Oh, I see how this is very confusing. `fe_var_id` refers to the ID of variables inserted in the frontend, ie. during the instrumentation injection. But this is not entirely correct since monomorphization is still part of the frontend. Long story short, this is different from the `var_id` because for generic functions we will end up with multiple instances of the same variable, all with different types. Ie. one `fe_var_id` corresponds to one or more `var_id`. `var_id` is the instantiated variable for which we keep track of the value. I'll think of better names.

Reply:
While changing all `u32` to more appropriate types, I renamed quite a few things. `fe_var_id` is now called `source_var_id` to convey the fact that it refers to a variable found in the source code, while `var_id` (or `debug_var_id`) is used to name the runtime variable which has a unique type, determined by the monomorphizer.